### ISSUE TITLE
doc_markdown false positive with ref link names

### ISSUE THREAD

**Issue by killercup** [Manishearth/rust-clippy#883]
To [this](https://github.com/diesel-rs/diesel/blob/9c5fec63c081ac14ed920f18bb6216d6b786f0a7/diesel/src/query_dsl/boxed_dsl.rs#L60-L67) doc comment, clippy says:

> warning: you should put `helper_types][helper_types` between ticks in the documentation, #[warn(doc_markdown)] on by default
>
> […]
>
> warning: you should put `helper_types` between ticks in the documentation, #[warn(doc_markdown)] on by default

It wants me to put _both_ occurrences of 'helper_types' in back ticks, and doesn't realize the second one is actually the named link references (or whatever this is called in markdown).

Using a small regex, we can remove these false positives, but will probably introduce some false negatives. A better but more complicated option would be to introduce a markdown parser/special markdown linter to clippy. (Or move the whole doc comment linting into a separate lint crate as suggested earlier.)

**Comment by mcarton** [comment]
AFAICT it is working as wanted. It should be:
```
The [`helper_types`][helper_types] module

[helper_types]: ../helper_types/index.html
```

to be rendered:

> The [`helper_types`][helper_types] module

[helper_types]: ../helper_types/index.html

the name of the module being some piece of code, it should be in ticks.

**Comment by killercup** [comment]
Sorry, I should've made this clearer. I changed to code locally to 

```rust
/// function. The [`helper_types`][helper_types] module exists to help with this,
```

and it still wants me to "put `helper_types` between ticks in the documentation". To me, this means the lint is talking about the _second_ occurrence of "helper_types" which should not be in back ticks.

**Comment by mcarton** [comment]
Arf :cry:

It’s weird, there is some test for that:
```rust
/// And here is another [inline link][inline_link].
/// It can also be [inline_link2].
```
[[source]](https://github.com/Manishearth/rust-clippy/blob/master/tests/compile-fail/doc.rs#L88)

**Comment by mcarton** [comment]
@killercup are you sure about that? I can’t reproduce.

**Comment by llogiq** [comment]
@killercup, perhaps you use an older clippy version?

**Comment by killercup** [comment]
Congratulations! It seems to have been fixed in 0.0.64 (which diesel can use since the last nightly bump) 😃 

If anyone wants to see some of the lint warnings diesel currently gets: https://travis-ci.org/diesel-rs/diesel/jobs/126730544#L1039

**Comment by mcarton** [comment]
\o/
